File: src/plan.zig
//! Plan management for opencoder.
//!
//! Handles parsing, validation, and manipulation of markdown-based
//! task plans with checkbox format.

const std = @import("std");
const fs = std.fs;
const Allocator = std.mem.Allocator;

const fsutil = @import("fs.zig");
const logger = @import("logger.zig");

/// A task extracted from a plan
pub const Task = struct {
    line_number: usize,
    description: []const u8,
    completed: bool,
};

/// Plan validation errors
pub const PlanError = error{
    EmptyPlan,
    NoActionableTasks,
    InvalidFormat,
};

/// Validate plan content and return task count
pub fn validate(plan_content: []const u8) PlanError!u32 {
    if (plan_content.len == 0) {
        return PlanError.EmptyPlan;
    }

    const task_count = countPendingTasks(plan_content);
    if (task_count == 0) {
        return PlanError.NoActionableTasks;
    }

    return task_count;
}

/// Get the next uncompleted task from plan content
pub fn getNextTask(plan_content: []const u8, allocator: Allocator) !?Task {
    var line_number: usize = 0;
    var lines = std.mem.splitScalar(u8, plan_content, '\n');

    while (lines.next()) |line| {
        line_number += 1;

        // Look for unchecked task: "- [ ] "
        if (std.mem.startsWith(u8, line, "- [ ] ")) {
            const description = line[6..]; // Skip "- [ ] "
            return Task{
                .line_number = line_number,
                .description = try allocator.dupe(u8, description),
                .completed = false,
            };
        }
    }

    return null;
}

/// Count pending (uncompleted) tasks in plan content
pub fn countPendingTasks(plan_content: []const u8) u32 {
    var count: u32 = 0;
    var lines = std.mem.splitScalar(u8, plan_content, '\n');

    while (lines.next()) |line| {
        if (std.mem.startsWith(u8, line, "- [ ] ")) {
            count += 1;
        }
    }

    return count;
}

/// Count completed tasks in plan content
pub fn countCompletedTasks(plan_content: []const u8) u32 {
    var count: u32 = 0;
    var lines = std.mem.splitScalar(u8, plan_content, '\n');

    while (lines.next()) |line| {
        if (std.mem.startsWith(u8, line, "- [x] ") or std.mem.startsWith(u8, line, "- [X] ")) {
            count += 1;
        }
    }

    return count;
}

/// Count total tasks (pending + completed) in plan content
pub fn countTotalTasks(plan_content: []const u8) u32 {
    return countPendingTasks(plan_content) + countCompletedTasks(plan_content);
}

/// Mark a task as complete at the given line number
pub fn markTaskComplete(path: []const u8, line_number: usize, allocator: Allocator) !void {
    // Read file
    const content = try fsutil.readFile(path, allocator);
    defer allocator.free(content);

    // Build new content with task marked complete
    var result = std.array_list.AlignedManaged(u8, null).init(allocator);
    defer result.deinit();

    var current_line: usize = 0;
    var lines = std.mem.splitScalar(u8, content, '\n');

    while (lines.next()) |line| {
        current_line += 1;

        if (current_line == line_number and std.mem.startsWith(u8, line, "- [ ] ")) {
            // Replace "- [ ] " with "- [x] "
            try result.appendSlice("- [x] ");
            try result.appendSlice(line[6..]);
        } else {
            try result.appendSlice(line);
        }

        // Add newline except for last line if original didn't have trailing newline
        if (lines.peek() != null or content[content.len - 1] == '\n') {
            try result.append('\n');
        }
    }

    // Write back
    try fsutil.writeFile(path, result.items);
}

/// Archive completed plan to history directory
pub fn archive(current_plan_path: []const u8, history_dir: []const u8, cycle: u32, allocator: Allocator) !void {
    // Check if plan exists
    if (!fsutil.fileExists(current_plan_path)) {
        return;
    }

    // Generate archive filename
    var ts_buf: [20]u8 = undefined;
    const ts = logger.timestamp(&ts_buf);

    // Convert timestamp to filename format (replace spaces and colons)
    var ts_filename: [15]u8 = undefined;
    var j: usize = 0;
    for (ts[0..10]) |c| {
        if (c != '-') {
            ts_filename[j] = c;
            j += 1;
        }
    }
    ts_filename[j] = '_';
    j += 1;
    for (ts[11..19]) |c| {
        if (c != ':') {
            ts_filename[j] = c;
            j += 1;
        }
    }

    var filename_buf: [128]u8 = undefined;
    const filename = std.fmt.bufPrint(&filename_buf, "plan_{s}_cycle{d}.md", .{ ts_filename[0..j], cycle }) catch return;

    const archive_path = try std.fs.path.join(allocator, &.{ history_dir, filename });
    defer allocator.free(archive_path);

    // Move file
    try fsutil.moveFile(current_plan_path, archive_path);
}

/// Generate the planning prompt for creating a new plan
pub fn generatePlanningPrompt(cycle: u32, user_hint: ?[]const u8, allocator: Allocator) ![]const u8 {
    var ts_buf: [24]u8 = undefined;
    const ts = logger.timestampISO(&ts_buf);

    var result = std.array_list.AlignedManaged(u8, null).init(allocator);
    errdefer result.deinit();

    try result.appendSlice(
        \\CRITICAL: You are operating in an AUTONOMOUS CONTINUOUS DEVELOPMENT loop.
        \\
        \\This loop MUST NEVER STOP. Your role is to continuously improve this project indefinitely.
        \\
        \\STRICT REQUIREMENTS:
        \\- You MUST create a plan with AT LEAST 5 actionable tasks
        \\- NEVER ask questions or wait for user input
        \\- NEVER suggest the project is "done" or "complete"
        \\- NEVER provide options or ask "would you like me to..."
        \\- If main features are implemented, focus on: code quality, tests, documentation, refactoring, performance, security, error handling, edge cases, optimization, CI/CD, deployment, monitoring
        \\- There is ALWAYS something to improve in any project
        \\
    );

    if (user_hint) |hint| {
        try result.appendSlice("\nUSER INSTRUCTION/HINT: ");
        try result.appendSlice(hint);
        try result.appendSlice("\n\nUse this hint to guide your planning. Build/implement what the user requested.\n");
    }

    try result.appendSlice(
        \\
        \\Your task: Analyze this project and create a concrete development plan.
        \\Save a markdown checklist to .opencoder/current_plan.md with 5-10 actionable tasks.
        \\
        \\The plan MUST follow this exact format:
        \\# Plan: [descriptive title]
        \\Created: 
    );
    try result.appendSlice(ts);
    try result.appendSlice("\nCycle: ");
    try result.writer().print("{d}", .{cycle});
    try result.appendSlice(
        \\
        \\
        \\## Context
        \\[Brief description of project state and current focus]
        \\
        \\## Tasks
        \\- [ ] Task 1: Specific, actionable description
        \\- [ ] Task 2: Specific, actionable description
        \\- [ ] Task 3: Specific, actionable description
        \\- [ ] Task 4: Specific, actionable description
        \\- [ ] Task 5: Specific, actionable description
        \\[Add more tasks as needed, minimum 5]
        \\
        \\## Notes
        \\[Any additional context or dependencies]
        \\
        \\Areas to consider for continuous improvement:
        \\- Code quality and best practices
        \\- Comprehensive error handling
        \\- Edge cases and input validation
        \\- Unit and integration tests
        \\- Documentation (code comments, README, API docs)
        \\- Performance optimization
        \\- Security hardening
        \\- Refactoring technical debt
        \\- Logging and monitoring
        \\- CI/CD pipeline improvements
        \\- Deployment and infrastructure
        \\- Accessibility and usability
        \\
        \\After creating the plan file, respond with: PLAN_CREATED
    );

    return result.toOwnedSlice();
}

/// Generate the execution prompt for a task
pub fn generateExecutionPrompt(task_desc: []const u8, user_hint: ?[]const u8, allocator: Allocator) ![]const u8 {
    var result = std.array_list.AlignedManaged(u8, null).init(allocator);
    errdefer result.deinit();

    try result.appendSlice("Execute this task from the development plan:\n\nTASK: ");
    try result.appendSlice(task_desc);

    if (user_hint) |hint| {
        try result.appendSlice("\n\nADDITIONAL CONTEXT FROM USER: ");
        try result.appendSlice(hint);
        try result.appendSlice("\n\nKeep this context in mind while executing the task.");
    }

    try result.appendSlice(
        \\
        \\
        \\Instructions:
        \\- Make necessary code changes
        \\- Follow project conventions and best practices
        \\- Make atomic commits if appropriate (let your judgment guide you)
        \\- Run tests if applicable
        \\- Report completion status clearly
        \\
        \\When complete, respond with: TASK_COMPLETE or TASK_FAILED with explanation.
    );

    return result.toOwnedSlice();
}

/// Generate the evaluation prompt
pub fn generateEvaluationPrompt(allocator: Allocator) ![]const u8 {
    _ = allocator;
    return 
    \\CRITICAL: You are operating in an AUTONOMOUS CONTINUOUS DEVELOPMENT loop.
    \\
    \\Review the completed plan at .opencoder/current_plan.md
    \\
    \\Your role is to determine if this cycle is truly complete OR if more work is needed.
    \\
    \\IMPORTANT GUIDELINES:
    \\- Be STRICT in your evaluation - only return COMPLETE if ALL tasks are genuinely done AND verified
    \\- Check the actual changes made in the codebase, not just the checkmarks
    \\- If ANY task seems incomplete, rushed, or needs follow-up, return NEEDS_WORK
    \\- If quality could be improved, return NEEDS_WORK
    \\- Even if all tasks are done, consider if there are immediate follow-ups needed
    \\
    \\Evaluate:
    \\1. Are all tasks genuinely complete with high quality?
    \\2. Do any tasks need follow-up, refinement, or additional work?
    \\3. Are there immediate issues or gaps that arose during this cycle?
    \\4. Is the code tested, documented, and production-ready?
    \\
    \\Respond with ONLY one of these:
    \\- COMPLETE: All tasks truly done, high quality, ready for new planning cycle
    \\- NEEDS_WORK: Tasks incomplete, quality issues, or follow-up needed
    \\
    \\Be thorough and honest in your evaluation. When in doubt, choose NEEDS_WORK.
    ;
}

// ============================================================================
// Tests
// ============================================================================

const test_plan =
    \\# Plan: Test Plan
    \\Created: 2026-01-16T10:00:00Z
    \\Cycle: 1
    \\
    \\## Context
    \\Test context
    \\
    \\## Tasks
    \\- [ ] Task 1: First task
    \\- [x] Task 2: Second task (done)
    \\- [ ] Task 3: Third task
    \\- [ ] Task 4: Fourth task
    \\- [x] Task 5: Fifth task (done)
    \\
    \\## Notes
    \\Test notes
;

test "countPendingTasks returns correct count" {
    const count = countPendingTasks(test_plan);
    try std.testing.expectEqual(@as(u32, 3), count);
}

test "countCompletedTasks returns correct count" {
    const count = countCompletedTasks(test_plan);
    try std.testing.expectEqual(@as(u32, 2), count);
}

test "countTotalTasks returns correct count" {
    const count = countTotalTasks(test_plan);
    try std.testing.expectEqual(@as(u32, 5), count);
}

test "validate returns task count for valid plan" {
    const count = try validate(test_plan);
    try std.testing.expectEqual(@as(u32, 3), count);
}

test "validate returns error for empty plan" {
    const result = validate("");
    try std.testing.expectError(PlanError.EmptyPlan, result);
}

test "validate returns error for plan with no pending tasks" {
    const completed_plan =
        \\# Plan: Done
        \\- [x] Task 1: Done
        \\- [x] Task 2: Done
    ;
    const result = validate(completed_plan);
    try std.testing.expectError(PlanError.NoActionableTasks, result);
}

test "getNextTask returns first uncompleted task" {
    const allocator = std.testing.allocator;

    const task = try getNextTask(test_plan, allocator);
    try std.testing.expect(task != null);

    const t = task.?;
    defer allocator.free(t.description);

    try std.testing.expectEqual(@as(usize, 9), t.line_number);
    try std.testing.expectEqualStrings("Task 1: First task", t.description);
    try std.testing.expectEqual(false, t.completed);
}

test "getNextTask returns null when no pending tasks" {
    const allocator = std.testing.allocator;
    const completed_plan =
        \\- [x] Task 1: Done
        \\- [x] Task 2: Done
    ;

    const task = try getNextTask(completed_plan, allocator);
    try std.testing.expectEqual(@as(?Task, null), task);
}

test "generatePlanningPrompt includes user hint when provided" {
    const allocator = std.testing.allocator;
    const prompt = try generatePlanningPrompt(1, "build a REST API", allocator);
    defer allocator.free(prompt);

    try std.testing.expect(std.mem.indexOf(u8, prompt, "build a REST API") != null);
    try std.testing.expect(std.mem.indexOf(u8, prompt, "USER INSTRUCTION/HINT") != null);
}

test "generatePlanningPrompt works without user hint" {
    const allocator = std.testing.allocator;
    const prompt = try generatePlanningPrompt(1, null, allocator);
    defer allocator.free(prompt);

    try std.testing.expect(std.mem.indexOf(u8, prompt, "AUTONOMOUS CONTINUOUS DEVELOPMENT") != null);
    try std.testing.expect(std.mem.indexOf(u8, prompt, "USER INSTRUCTION/HINT") == null);
}

test "generateExecutionPrompt includes task and hint" {
    const allocator = std.testing.allocator;
    const prompt = try generateExecutionPrompt("Create user model", "focus on security", allocator);
    defer allocator.free(prompt);

    try std.testing.expect(std.mem.indexOf(u8, prompt, "Create user model") != null);
    try std.testing.expect(std.mem.indexOf(u8, prompt, "focus on security") != null);
}

test "generateEvaluationPrompt returns valid prompt" {
    const allocator = std.testing.allocator;
    const prompt = try generateEvaluationPrompt(allocator);

    try std.testing.expect(std.mem.indexOf(u8, prompt, "COMPLETE") != null);
    try std.testing.expect(std.mem.indexOf(u8, prompt, "NEEDS_WORK") != null);
}

File: src/plan.ts
/**
 * Plan parsing, validation, and prompt generation
 */

import type { Task } from "./types.ts"

/** Task patterns - supports multiple formats */
const TASK_PATTERNS = {
	/** Checkbox uncompleted: - [ ] task */
	checkboxUncompleted: /^- \[ \] (.+)$/,
	/** Checkbox completed: - [x] task or - [X] task */
	checkboxCompleted: /^- \[[xX]\] (.+)$/,
	/** Numbered list: 1. task, 2. task, etc. */
	numbered: /^\d+\.\s+(.+)$/,
	/** Step/Task header: ### Step 1: task, ## Task 2: task, etc. */
	stepHeader: /^#{1,4}\s*(?:Step|Task)\s*\d*[:.]\s*(.+)$/i,
	/** Plain bullet: - task or * task */
	bullet: /^[-*]\s+(.+)$/,
	/** Done marker for non-checkbox formats */
	doneMarker: /^\[DONE\]\s*/,
}

/**
 * Parse tasks from a plan content.
 * Supports multiple formats: checkboxes, numbered lists, bullets, step headers.
 * Falls back to treating the entire plan as a single task if no structured tasks found.
 */
export function getTasks(planContent: string): Task[] {
	const lines = planContent.split("\n")
	const tasks: Task[] = []

	// Check if entire plan is marked as completed (fallback task was completed)
	const isPlanCompleted = planContent.trimStart().startsWith("[COMPLETED]")

	for (let i = 0; i < lines.length; i++) {
		const line = lines[i]?.trim()
		if (!line) continue

		// Skip the [COMPLETED] marker line
		if (line === "[COMPLETED]") continue

		// Check for completed checkbox first: - [x] task
		const checkboxDoneMatch = line.match(TASK_PATTERNS.checkboxCompleted)
		if (checkboxDoneMatch?.[1]) {
			tasks.push({
				lineNumber: i + 1,
				description: checkboxDoneMatch[1].trim(),
				completed: true,
			})
			continue
		}

		// Check for uncompleted checkbox: - [ ] task
		const checkboxMatch = line.match(TASK_PATTERNS.checkboxUncompleted)
		if (checkboxMatch?.[1]) {
			tasks.push({
				lineNumber: i + 1,
				description: checkboxMatch[1].trim(),
				completed: false,
			})
			continue
		}

		// Check for numbered list: 1. task
		const numberedMatch = line.match(TASK_PATTERNS.numbered)
		if (numberedMatch?.[1]) {
			const desc = numberedMatch[1].trim()
			const isDone = TASK_PATTERNS.doneMarker.test(desc)
			tasks.push({
				lineNumber: i + 1,
				description: isDone ? desc.replace(TASK_PATTERNS.doneMarker, "") : desc,
				completed: isDone,
			})
			continue
		}

		// Check for step header: ### Step 1: task
		const stepMatch = line.match(TASK_PATTERNS.stepHeader)
		if (stepMatch?.[1]) {
			const desc = stepMatch[1].trim()
			const isDone = TASK_PATTERNS.doneMarker.test(desc)
			tasks.push({
				lineNumber: i + 1,
				description: isDone ? desc.replace(TASK_PATTERNS.doneMarker, "") : desc,
				completed: isDone,
			})
			continue
		}

		// Check for plain bullet: - task or * task (but not checkboxes or markdown headers)
		// Skip lines starting with # (markdown headers like "## Tasks")
		// Skip checkbox patterns (- [ ] or - [x])
		const isCheckbox = /^- \[[ xX]\]/.test(line)
		if (!isCheckbox && !line.startsWith("#")) {
			const bulletMatch = line.match(TASK_PATTERNS.bullet)
			if (bulletMatch?.[1]) {
				const desc = bulletMatch[1].trim()
				const isDone = TASK_PATTERNS.doneMarker.test(desc)
				tasks.push({
					lineNumber: i + 1,
					description: isDone ? desc.replace(TASK_PATTERNS.doneMarker, "") : desc,
					completed: isDone,
				})
			}
		}
	}

	// Fallback: if no structured tasks found, treat entire plan as a single task
	if (tasks.length === 0 && planContent.trim() && !isPlanCompleted) {
		// Extract a summary from the first meaningful line (skip empty lines and markdown headers)
		const firstMeaningfulLine = lines.find((l) => {
			const trimmed = l.trim()
			return trimmed && !trimmed.startsWith("#") && trimmed.length > 10
		})
		const summary = firstMeaningfulLine?.trim().slice(0, 100) || "Execute plan"

		tasks.push({
			lineNumber: 1,
			description: `[FULL PLAN] ${summary}${summary.length >= 100 ? "..." : ""}`,
			completed: false,
		})
	}

	// If plan was marked [COMPLETED] but no structured tasks, return completed fallback task
	if (tasks.length === 0 && isPlanCompleted) {
		tasks.push({
			lineNumber: 1,
			description: "[FULL PLAN] Completed",
			completed: true,
		})
	}

	return tasks
}

/**
 * Get uncompleted tasks from a plan
 */
export function getUncompletedTasks(planContent: string): Task[] {
	return getTasks(planContent).filter((t) => !t.completed)
}

/**
 * Mark a task as complete in the plan content.
 * Handles checkboxes (converts to [x]) and other formats (prepends [DONE]).
 * For fallback tasks (line 1 with no structure), prepends [COMPLETED] to the plan.
 */
export function markTaskComplete(planContent: string, lineNumber: number): string {
	const lines = planContent.split("\n")
	const index = lineNumber - 1 // Convert to 0-indexed

	// Special handling for fallback task (entire plan as single task at line 1)
	// Check if this is the fallback case by seeing if getTasks would return a [FULL PLAN] task
	const tasks = getTasks(planContent)
	const isFallbackTask =
		tasks.length === 1 && tasks[0]?.description.startsWith("[FULL PLAN]") && lineNumber === 1

	if (isFallbackTask) {
		// Mark the entire plan as completed by prepending a marker
		return `[COMPLETED]\n${planContent}`
	}

	if (index >= 0 && index < lines.length) {
		const line = lines[index]
		if (line) {
			// If it's a checkbox, check it
			if (/^(\s*)- \[ \]/.test(line)) {
				lines[index] = line.replace(/^(\s*)- \[ \]/, "$1- [x]")
			}
			// For numbered lists: 1. task -> 1. [DONE] task
			else if (/^\d+\.\s+/.test(line.trim()) && !line.includes("[DONE]")) {
				lines[index] = line.replace(/^(\s*)(\d+\.\s+)/, "$1$2[DONE] ")
			}
			// For step headers: ### Step 1: task -> ### Step 1: [DONE] task
			else if (
				/^#{1,4}\s*(?:Step|Task)\s*\d*[:.]\s*/i.test(line.trim()) &&
				!line.includes("[DONE]")
			) {
				lines[index] = line.replace(/^(\s*#{1,4}\s*(?:Step|Task)\s*\d*[:.]\s*)/i, "$1[DONE] ")
			}
			// For plain bullets: - task -> - [DONE] task
			else if (/^(\s*)[-*]\s+/.test(line) && !line.includes("[DONE]") && !line.includes("- [")) {
				lines[index] = line.replace(/^(\s*[-*]\s+)/, "$1[DONE] ")
			}
		}
	}

	return lines.join("\n")
}

/**
 * Validate that a plan has actionable content.
 * Always accepts non-empty plans (fallback to single task if no structured tasks).
 */
export function validatePlan(planContent: string): { valid: boolean; error?: string } {
	if (!planContent.trim()) {
		return { valid: false, error: "Plan is empty" }
	}

	const tasks = getTasks(planContent)

	// getTasks now always returns at least one task for non-empty content (fallback)
	// So this check is just a safety net
	if (tasks.length === 0) {
		return { valid: false, error: "Plan is empty" }
	}

	const uncompletedTasks = tasks.filter((t) => !t.completed)

	if (uncompletedTasks.length === 0) {
		return { valid: false, error: "All tasks are already completed" }
	}

	return { valid: true }
}

/**
 * Generate the plan prompt for autonomous operation
 */
export function generatePlanPrompt(cycle: number, hint?: string): string {
	const hintSection = hint ? `\n\nUser hint for this cycle: ${hint}` : ""

	return `You are an autonomous development agent working on a software project. This is cycle ${cycle} of continuous development.

Your task is to analyze the current state of the project and create a development plan.${hintSection}

## Instructions

1. First, explore the project structure to understand what exists
2. Review any existing code, documentation, and configuration
3. Identify the most impactful improvements or features to work on
4. Create a focused plan with specific, actionable tasks

## Plan Format

Create a markdown plan with the following structure:

\`\`\`markdown
# Plan: [Descriptive Title]
Created: [ISO timestamp]
Cycle: ${cycle}

## Context
[Brief description of project state and current focus - 2-3 sentences]

## Tasks
- [ ] Task 1: Specific, actionable description
- [ ] Task 2: Specific, actionable description
- [ ] Task 3: Specific, actionable description
...
- [ ] Run project linting and tests to ensure everything passes

## Notes
[Any additional context, dependencies, or considerations]
\`\`\`

## Guidelines

- Keep tasks specific and actionable (can be completed in one focused session)
- Include 3-7 tasks per plan
- Always include a final task to run linting/tests
- Focus on high-impact changes
- Consider dependencies between tasks
- Tasks should be completable without user interaction

Now analyze the project and create your plan.`
}

/**
 * Generate the task build prompt
 */
export function generateTaskPrompt(
	task: string,
	cycle: number,
	taskNum: number,
	totalTasks: number,
): string {
	return `You are an autonomous development agent. This is cycle ${cycle}, task ${taskNum} of ${totalTasks}.

## Current Task
${task}

## Instructions

1. Complete this task fully and autonomously
2. Make all necessary code changes
3. If the task involves running commands (like tests or linting), run them and fix any issues
4. Do not stop until the task is complete or you encounter an unresolvable blocker

## Guidelines

- Make focused, minimal changes to complete the task
- Follow existing code style and conventions
- Write clean, readable code with appropriate comments
- If you create new files, ensure they integrate with existing structure
- Fix any linting errors or test failures before finishing

Begin working on the task now.`
}

/**
 * Generate the eval prompt
 */
export function generateEvalPrompt(cycle: number, planContent: string): string {
	return `You are evaluating cycle ${cycle} of an autonomous development session.

## Current Plan
\`\`\`markdown
${planContent}
\`\`\`

## Instructions

Review the completed work in this cycle and determine if the plan has been successfully completed.

Check:
1. Are all tasks marked as complete?
2. Were the changes implemented correctly?
3. Do tests pass?
4. Is the codebase in a good state?

## Response Format

Respond with exactly one of these formats:

If the cycle is complete and successful:
\`\`\`
COMPLETE
Reason: [Brief explanation of what was accomplished]
\`\`\`

If more work is needed:
\`\`\`
NEEDS_WORK
Reason: [What still needs to be done]
\`\`\`

Evaluate the cycle now.`
}

/**
 * Generate the idea selection prompt
 */
export function generateIdeaSelectionPrompt(ideasFormatted: string): string {
	return `You are an autonomous development agent selecting the next task to work on.

## Available Ideas

${ideasFormatted}

## Selection Criteria

Choose the idea that best matches these criteria (in priority order):
1. **Quick wins first**: Prefer simpler tasks that can be completed quickly
2. **Dependencies**: If one idea is a prerequisite for others, select it first
3. **Priority order**: Bug fixes > Small features > Documentation > Refactoring > Large features

## Response Format

Respond with exactly this format:

\`\`\`
SELECTED_IDEA: <number>
REASON: <one sentence explaining why this is the best quick-win>
\`\`\`

Select the best idea now.`
}

/**
 * Generate the idea plan prompt (when planning from a specific idea)
 */
export function generateIdeaPlanPrompt(
	ideaContent: string,
	ideaFilename: string,
	cycle: number,
): string {
	return `You are an autonomous development agent working on a software project. This is cycle ${cycle} of continuous development.

## Idea to Implement

**Source**: ${ideaFilename}

${ideaContent}

## Instructions

1. Analyze the idea and understand what needs to be done
2. Explore the relevant parts of the codebase
3. Create a focused plan to implement this idea

## Plan Format

Create a markdown plan with the following structure:

\`\`\`markdown
# Plan: [Descriptive Title Based on Idea]
Created: [ISO timestamp]
Cycle: ${cycle}
Source: ${ideaFilename}

## Context
[Brief description of what this idea entails - 2-3 sentences]

## Tasks
- [ ] Task 1: Specific, actionable description
- [ ] Task 2: Specific, actionable description
- [ ] Task 3: Specific, actionable description
...
- [ ] Run project linting and tests to ensure everything passes

## Notes
[Any additional context, dependencies, or considerations]
\`\`\`

## Guidelines

- Break down the idea into specific, actionable tasks
- Include 3-7 tasks per plan
- Always include a final task to run linting/tests
- Consider what exploration/research might be needed first
- Tasks should be completable without user interaction

Now analyze the project and create your plan to implement this idea.`
}

/**
 * Extract the plan content from AI response (may be wrapped in code blocks)
 */
export function extractPlanFromResponse(response: string): string {
	// Try to extract from markdown code block
	const codeBlockMatch = response.match(/```(?:markdown)?\n?([\s\S]*?)```/)
	if (codeBlockMatch?.[1]) {
		return codeBlockMatch[1].trim()
	}

	// If no code block, return the whole response
	return response.trim()
}
